### PR TITLE
Create SharedArray from disk file

### PR DESCRIPTION
**Pull request by timholy**
The third commit here allows you to share data from a pre-existing disk file (or a newly-created one) across all workers.

The first commit is just a cleanup (and net code deletion): it leverages #10525 for indexing, and makes all fields of a SharedArray concrete (I think the important ones were already). The second is #12423 without the documentation changes (because docs can always be added later, and we're in a bit of doc-paralysis at the moment). That commit should still be good even after this merges.

CC @amitmurthy

### CODE REVIEW

File: base/sharedarray.jl [hunk 1]
```diff
@@ -114,11 +100,75 @@ end
...
+
+    # If not supplied, determine the appropriate mode
+    have_file = onlocalhost ? isfile(filename) : remotecall_fetch(pids[1], isfile, filename)
+    if mode == nothing
+        mode = have_file ? "r+" : "w+"
+    end
+    workermode = mode == "w+" ? "r+" : mode  # workers don't truncate!
+
+    # Ensure the file will be readable
+    mode in ("r", "r+", "w+", "a+") || error("mode must be readable, but I got $mode")
```
Comment (jakebolewski) [inline base/sharedarray.jl]: can these be `ArgumentError`'s?  Also, using a personal pronoun in an error message is a bit weird.

Comment (timholy) [inline base/sharedarray.jl]: Good points all round, will fix.

**Comment by timholy** [comment]
OK, the remaining AppVeyor error seems unrelated (`LoadError("C:\\projects\\julia\\test\\pollfd.jl",112,ErrorException("test failed: Any[] == [1]\n in expression: a == [1]"))`). The one oddity is that [the `mmap` code throws different error types on Windows vs Unix](https://github.com/JuliaLang/julia/blob/11efb31fe3d91bf9c2b654cc72106f8fb87c4d33/test/parallel.jl#L129-L135). Is this something we should harmonize, or is the split in testing fine?

**Comment by tkelman** [comment]
Where are the throws coming from? If it's really the same condition then it would make sense to me to make it the same exception type if that's not much work.

File: test/parallel.jl [hunk 2]
```diff
@@ -86,6 +97,66 @@ for p in procs(d)
     @test d[idxl] == p
 end
 
+### SharedArrays from a file
+
+# Mapping an existing file
+fn = tempname()
+open(fn, "w") do io
```
Comment (tkelman) [inline test/parallel.jl]: these should probably clean up after themselves when they're done

Comment (timholy) [inline test/parallel.jl]: Fixed in 13846e5d14649ac69992ed6cfd8c89c255d0f2f2

**Pull request merged**